### ISSUE TITLE
Enhance DynamicallyAccessedMembers Attribute to apply to class 

### ISSUE THREAD

**Issue by LakshanF** [dotnet/runtime#49465]
## Background and Motivation
This proposal is taking @MichalStrehovsky's trimer [proposal ](https://github.com/mono/linker/issues/1196)through its API change part. This proposal aligns well with the API proposal #49448 in enforcing the type that are required to be preserved for all derived types of a base class.  #49448 helps with validating the type tree (parent type and its descendants) and this API enhancement will help with the enforcement.

The `DynamicallyAccessedMembersAttribute `[attribute ](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute) currently cannot be applied to a class or interface. Adding this support will allow developers to annotate a type to help the trimmer preserve its members that are being used. If the trimmer needs to keep the type, it will keep the required members on its derived types as well.

## Proposed API

```diff
 namespace System.Diagnostics.CodeAnalysis
 {
     [AttributeUsage(
+        AttributeTargets.Class |
         AttributeTargets.Field |
         AttributeTargets.GenericParameter |
+        AttributeTargets.Interface |
+        AttributeTargets.Struct |
         AttributeTargets.Method |
         AttributeTargets.Parameter |
         AttributeTargets.Property |
         AttributeTargets.ReturnValue,
         Inherited = false)]
     public partial class DynamicallyAccessedMembersAttribute : Attribute
     {
     }
 }
```

## Usage Examples
For a concrete example, see the existing [code ](https://github.com/dotnet/runtime/blob/267cb42ba046bf727dff41299322fbf5e4ac27fa/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L2768) in `EventSources `which will create a manifest for a derived `EventSource`,
```
                         m_rawManifest = CreateManifestAndDescriptors(this.GetType(), Name, this);
```
If the EventSource Type is annotated with DynamicallyAccessedMembersAttribute as shown below, then the trimmer can preserve all methods of the derived EventSource and create a correct manifest for the derived EventSource type. 
```
[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
public class EventSource
```
Without such support, some event methods might not be included in the manifest generation, which can cause tool problems among other issue. See [this ](https://github.com/dotnet/runtime/issues/32863#issuecomment-755963798)customer comment for more details.

## Alternative Designs
The API proposal  #49448 on its own

## Risks
This API might cause more code to be preserved than intended since it applies to all derived types of the annotated type.

**Comment by LakshanF** [comment]
cc @vitek-karas, @MichalStrehovsky

**Comment by terrajobst** [comment]
* Looks good as proposed

```diff
namespace System.Diagnostics.CodeAnalysis
 {
     [AttributeUsage(
+        AttributeTargets.Class |
         AttributeTargets.Field |
         AttributeTargets.GenericParameter |
+        AttributeTargets.Interface |
         AttributeTargets.Method |
         AttributeTargets.Parameter |
         AttributeTargets.Property |
         AttributeTargets.ReturnValue,
         Inherited = false)]
     public partial class DynamicallyAccessedMembersAttribute : Attribute
     {
     }
 }
 ```

**Comment by eerhardt** [comment]
Resolved by #49778.
